### ISSUE TITLE
Gitea mirroring security requirements 

### ISSUE THREAD

**Issue by JimMadge** [alan-turing-institute/data-safe-haven#1995]
We are aiming to use the internal Gitea instance to mirror external git repositories in a read only manner.

We need to investigate how to do this in a secure way that,

- Does not allow data egreess (_e.g._ allowing pushing to external git repositories)
- Enables admins to review/approve/revoke mirrors (_e.g._ we are probably not comfortable with users being able to add access to arbitrary repositories themselves)

We should design an architecture/configuration that supports these requirements.

**Comment by JimMadge** [comment]
Gitea can set up [pull repositories](https://docs.gitea.com/next/usage/repo-mirror#pulling-from-a-remote-repository) where you are not able to push back to the "original" repository.

The repository is periodically synced and can be synced ad-hoc

- How often is periodic sync? Can we configure that?
- _Who_ can start an ad-hoc sync?

**Comment by JimMadge** [comment]
With Gitea is is possible to enable browsing repositories without signing in (this is enabled by default).

This means we could have a Gitea instance which users are not expected to authenticate with, to browse the read only mirrors.
(I assume that would mean an additional instance, which might be a little confusing, but we could use sensible domain names/theming to differentiate them)

An authenticated user can create mirrors

<img width="851" alt="Screenshot 2024-07-08 at 15 22 46" src="https://github.com/alan-turing-institute/data-safe-haven/assets/23616154/416a98e9-1e02-43fb-83c0-294bd90126c8">

When not authenticated, you can only browse

<img width="851" alt="Screenshot 2024-07-08 at 15 21 31" src="https://github.com/alan-turing-institute/data-safe-haven/assets/23616154/f2f63f79-2435-4c5e-8269-65ac48b688e5">

Configuration,

- Disable self-registration: True
- Show register button: False
- Enable OpenID self-regirstration/authentication: False
- Require sign-in to view pages: False

**Comment by JimMadge** [comment]
@craddm @jemrobinson does a two Gitea instance solution (one for internal development, one for mirrors) sounds good to you?

**Comment by craddm** [comment]
Yes, and giving them sensible domain names sounds good (could be as simple as `external.*` and `internal.*`)

**Comment by jemrobinson** [comment]
Can you stop a non-authenticated user from registering? I wonder whether a simpler solution is to do what we used to do with package mirrors:
- external server: able to access the internet, not accessible by users
- internal server: not able to access the internet, accessible by users

with mirroring rules set up from internal to external AND external to GitHub/wherever.

**Comment by JimMadge** [comment]
> Can you stop a non-authenticated user from registering?

Yes, you can disable registration methods so that only an admin could create new users. Similar to how Nexus works.

I think this idea is simpler, less networking configuration and you would only need to create one repository migration.

**Comment by jemrobinson** [comment]
I'm still a bit nervous about allowing the user-accessible Gitea server to have unrestricted outgoing access to GitHub/GitLab/whatever. How can we convince ourselves that users couldn't (a) get admin rights on this and/or (b) use it as an egress pathway?

**Comment by JimMadge** [comment]
Yes, I think those are good questions. If we aren't confident, then something like an airgaped set of Gitea instances should work.

Luckily, I thin a lot of the infrastructure we need is the same in either case. We still want two Gitea instances and a method for creating read only mirrors to an instance inside an SRE.

AFAIK, what I proposed would give users inside the TRE no way to register an account. They could however authenticate _if_ they had the credentials of a valid account. _If_ they did that, I think it would be an easy egress pathway as you could create a push mirror to an external repository.

**Comment by jemrobinson** [comment]
@JimMadge @craddm : Would something like [copybara](https://github.com/google/copybara) running as either a container or a Function remove the need for two Gitea instances?

Or [git-mirror](https://github.com/beefsack/git-mirror) as a possibly simpler alternative?

**Comment by martintoreilly** [comment]
Outline architecture diagram from co-design discussion on 20 May ([source](https://drive.google.com/file/d/1PJa4Vo0h3DlkaY1PoB6ehjqOYDX9KWsp/view?usp=sharing))

**Key security requirements:**
- Airlocked mirror cannot be connect **to** from either outside or inside the SRE but can connect to:
    - GitHub (restricted via firewall), to pull updates from mirrored repositories
    - Internal user-facing Gitea instance (restricted via NSG), to push updates to mirrored repositories
- Internal user-facing Gitea mirror
    - This is the same instance as currently deployed and cannot access anything outside the SRE 

<img width="703" alt="Image" src="https://github.com/user-attachments/assets/7bdef108-7779-47e2-bff9-fb8778c0b641" />

**Comment by cptanalatriste** [comment]
From today's co-design session with @martintoreilly , @craddm , and @triangle-man : 

## Components 
- We will have two internal Gitea mirrors:- 1) An internal facing mirror accessible to users, that can't contact outside, and 2) an external facing mirror, inaccessible to users that can contact Github and the internal facing mirror.
- We will use a mechanism similar to our current `allowlist` to store the list of GitHub repositories to mirror. It will be an independent container that reads the allowlist and reconfigures Gitea.

## Networking
-  The firewall will only allow the external-facing mirror to make outbound connections to GitHub.
- Using network security groups rules, the external facing mirror can connect only to the internal facing mirror.

## User experience
- We will rely on Gitea's interval for keeping the mirror updated.
- The code developed inside the SRE can only be egressed following our current process (forms and approvals). It's not feasible to review entire repositories, but we can take a look to folders/files.
- Every SRE deployment, the following should happen: 1) Clean-up existing mirrors, 2) Per each mirror in the allowlist, we add a Pull Mirror to pull updates from an external repository (GitHub), 3) Per each Pul Mirror, we add a Push Mirror to push updates to the internal Gitea instance, 3) These mirrors should synchronise at least every 5 minutes.

## Implementation
Some useful resources from the Gitea API:

- Add a Push mirror to the repository: https://docs.gitea.com/api/1.22/#tag/repository/operation/repoAddPushMirror
- Sync a mirrored repository: https://docs.gitea.com/api/1.22/#tag/repository/operation/repoMirrorSync
- Explore what GitHub API exposes for making synchronisation smoother.


## Nice to haves
- Repositories updates under 5 minutes.
- User to trigger mirror updates from the SRE.
- Mirror updated when the original repo (Github) is updated.

**Comment by cptanalatriste** [comment]
Past related work, for reference: https://github.com/alan-turing-institute/data-safe-haven/pull/2071 and https://github.com/alan-turing-institute/data-safe-haven/pull/2008
